realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -884,6 +921,151 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAll(
+          Query* pQuery (SG(bgSharedGroupPtr)->import_from_handover(handoverQuery));
+          Table* pTable = pQuery->get_table().get();
+
+          if (!QUERY_VALID(env, pQuery) || !ROW_INDEXES_VALID(env, pTable, start, end, limit))
+              return -1;
+
+            // run the query
+            TableView* pResultView = new TableView( pQuery->find_all(S(start), S(end), S(limit)) );
+
+            // sorting the results
+            if (!COL_INDEX_VALID(env, pResultView, columnIndex))
+                      return -1;
+              int colType = pResultView->get_column_type( S(columnIndex) );
+
+              switch (colType) {
+                  case type_Bool:
+                  case type_Int:
+                  case type_DateTime:
+                  case type_Float:
+                  case type_Double:
+                  case type_String:
+                      pResultView->sort( S(columnIndex), ascending != 0 ? true : false);
+                      break;
+                  default:
+                      ThrowException(env, IllegalArgument, "Sort is currently only supported on integer, float, double, boolean, Date, and String columns.");
+                      return -1;
+              }
+
+            // handover the result
+            SharedGroup::Handover<TableView>* handover = SG(bgSharedGroupPtr)->export_for_handover(*pResultView, MutableSourcePayload::Move);
+
+            return reinterpret_cast<jlong>(handover);
+
+      } CATCH_STD()
+      return -1;
+  }
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllMultiSortedWithHandover
+  (JNIEnv *env, jobject, jlong bgSharedGroupPtr, jlong handoverQueryPtr, jlong start, jlong end, jlong limit, jlongArray columnIndices, jbooleanArray ascending)
+  {
+      try {
+          jsize arr_len = env->GetArrayLength(columnIndices);
+          jsize asc_len = env->GetArrayLength(ascending);
+
+          jlong *long_arr = env->GetLongArrayElements(columnIndices, NULL);
+          jboolean *bool_arr = env->GetBooleanArrayElements(ascending, NULL);
+
+          if (arr_len == 0) {
+              ThrowException(env, IllegalArgument, "You must provide at least one field name.");
+              return -1;
+          }
+          if (asc_len == 0) {
+              ThrowException(env, IllegalArgument, "You must provide at least one sort order.");
+              return -1;
+          }
+          if (arr_len != asc_len) {
+              ThrowException(env, IllegalArgument, "Number of column indices and sort orders do not match.");
+              return -1;
+          }
+
+           // cast the pointer
+           SharedGroup::Handover<Query>* handoverQuery = reinterpret_cast<SharedGroup::Handover<Query>*>(handoverQueryPtr);
+
+          // import the handover query pointer using the background SharedGroup
+          Query* pQuery (SG(bgSharedGroupPtr)->import_from_handover(handoverQuery));
+          Table* pTable = pQuery->get_table().get();
```

Comment:
We have not really used this naming convention before. I have no strong preferences, but we should keep it consistent.